internal/rpm/files.go

```diff
@@ -0,0 +1,149 @@
+package rpm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"maps"
+	"runtime"
+	"strings"
+	"sync"
+	"unique"
+	"weak"
+
+	"github.com/quay/zlog"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/quay/claircore"
+)
+
+// PkgCache is a process-wide cache for [PathSet]s.
+//
+// This allows for all callers to use a single [PathSet] for a given layer.
+var pkgCache fileCache
+
+// FileCache implements a cache of [PathSet] values keyed by the layer digest.
+type fileCache struct {
+	// map[unique.Handle[string]][weak.Pointer[PathSet]]
+	m  sync.Map
+	sf singleflight.Group
+}
+
+func (c *fileCache) cleanupFunc(wp weak.Pointer[PathSet]) func(unique.Handle[string]) {
+	return func(key unique.Handle[string]) {
+		c.m.CompareAndDelete(key, wp)
+	}
+}
+
+// GetPathSet returns a [PathSet] for the provided layer, using a cached
+// version if possible.
+func (c *fileCache) GetPathSet(ctx context.Context, layer *claircore.Layer) (*PathSet, error) {
+	// This makes use of [weak.Pointer], which may be odd at first.
+	//
+	// The weak package has full documentation, but roughly: because a
+	// [weak.Pointer] does not keep a value alive for GC purposes, every use
+	// that retrieves the original pointer needs to assume that it may have
+	// become nil and handle that case before dereferencing it.
+
+	key := unique.Make(layer.Hash.String())
+	if v, ok := c.m.Load(key); ok {
+		if s := v.(weak.Pointer[PathSet]).Value(); s != nil {
+			return s, nil
+		}
+	} else {
+		c.m.CompareAndDelete(key, v)
+	}
+
+	ch := c.sf.DoChan(key.Value(), func() (any, error) {
+		set := &PathSet{
+			paths: make(map[string]struct{}),
+		}
+		sys, err := layer.FS()
+		if err != nil {
+			return nil, fmt.Errorf("rpm: unable to open layer: %w", err)
+		}
+		seq, errFunc := FindDBs(ctx, sys)
+		defer func() {
+			err = errors.Join(err, errFunc())
+		}()
+
+		for found := range seq {
+			zlog.Debug(ctx).
+				Stringer("db", found).
+				Msg("found possible databases")
+			err = func() error {
+				db, err := OpenDB(ctx, sys, found)
+				if err != nil {
+					return err
+				}
+				defer db.Close()
+				ctx := zlog.ContextWithValues(ctx, "db", db.String())
+				zlog.Debug(ctx).Msg("examining database")
+				return db.populatePathSet(ctx, set)
+			}()
+			if err != nil {
+				return nil, err
+			}
+
+		}
+		wp := weak.Make(set)
+		runtime.AddCleanup(set, c.cleanupFunc(wp), key)
+		c.m.Store(key, wp)
+		return set, nil
+	})
+
+	select {
+	case <-ctx.Done():
+		return nil, context.Cause(ctx)
+	case res := <-ch:
+		if err := res.Err; err != nil {
+			return nil, err
+		}
+		return res.Val.(*PathSet), nil
+	}
+}
+
+// NewPathSet returns a [PathSet] for the provided layer.
+func NewPathSet(ctx context.Context, layer *claircore.Layer) (*PathSet, error) {
+	return pkgCache.GetPathSet(ctx, layer)
+}
+
+// PathSet is used to check if a path is an RPM-owned file.
+type PathSet struct {
+	_noCopy noCopy
```

Comment:
Curious why this cannot be copied

---

Is it to prevent concurrent access to `paths` and remove the need for locking?